### ISSUE TITLE
Dataset Distillation with Convexified Implicit Gradients

### ISSUE THREAD

**Issue by djamahl99** [Luoyadan/Research-Group-Progress#14]
# Background

## Data Distillation/Data Condensation

- Fairly new, still only working on small models/datasets due to gradient computation

- Early epoch models provide better signals for learning dataset distillation
- Can produce Membership Inference Attack robustness compared to models trained on full datasets

### Issues:
- Everyone is using different baselines to compare, each paper might pick 3 baselines and one might match :(
- Scaling up is difficult: e.g. takes large ratio for many methods to achieve full dataset performance (getting better)
- Starting to work on ImageNet, albeit usually at low resolutions / toy versions, but this paper does full resolution
- Often much more expensive to create these distilled datasets than it is to train on original dataset
- ViT, or anything larger than ResNet50 has often not been evaluated on the produced synthetic datasets (overfitting issue)

### SOTAs:
- Scaling Up Dataset Distillation to ImageNet-1K with Constant Memory (actually trains larger models e.g. ViT)
- Accelerating Dataset Distillation via Model Augmentation
- FRePo
- Dataset Distillation with Convexified Implicit Gradients

### Current research goals
- reduce memory consumption
- reduce computation cost
- larger models
- larger datasets

### Methods

- Gradient matching (DSA, IDC, FRePo)
- Trajectory matching (TESLA, DM, CAFE, MTT): e.g. match parameters between distilled and real data
- Parameterization (TM, RCIG)

### Dual problem
- Inner objective: update model parameters on the support / synthetic/distilled set
- Outer ovjective (meta): refine the distilled sets via meta-gradient updates

# Dataset Distillation with Convexified Implicit Gradients

Seems like the best method, but have only just discovered it so haven't read too far.

# Method

![image](https://user-images.githubusercontent.com/19633934/231649976-6b2a1cfd-602b-4706-948a-95918ae02770.png)

## Vanilla Implicit Gradients (VIG)
- implicit solution performs worse than random images, as implicit gradient theorem requires unique minimizer but deep neural networks are non-convex
- maximizes the mutial influence between our distilled dataset and the full dataset
- for deep models, influence functions are brittle and do not accurately estimate leave-one-out retraining. This discrepancy is partially because of the non-convexity of deep models

## Convexification (CIG)
- this paper aims to solve the problem with vanilla implicit gradients by making the model exhibit more convex behaviour
- they use the 1st-order Taylor approximation of learning dynamics, "linearized dynamics":
![image](https://user-images.githubusercontent.com/19633934/231648992-1777924a-9e58-455c-95b1-727a5c637fc6.png)
- if $\theta_0$ is fixed and we consider optimizing $\theta$, this new formulation is strongly convex in $\theta$ provided some L2 regularization
- this technique corresponds to learning on top of a frozen finite-width NTK
- we call this updated algorithm, CIG (convexified implicit gradients), which improves on VIG

## Combine Analytic Solutions and Implicit Gradients (RCIG)
- limitation of CIG: unable to find the true minima of the inner problem in any feasible amount of time , truncated gradient unrolls lead to bias in implicit gradients
- to perform better optimization of the inner problem - split network parameters into final layer and body, and not that, for any setting of the body parameters, we can ***efficiently analytically compute the optimal final layer*** (when trained with MSE)
![image](https://user-images.githubusercontent.com/19633934/231651668-18fcd99b-d51d-40d3-ba3a-513b1d890d97.png)
- given we can analytically solve for the final layer, we **only learn $\theta_b$** and automatically compute the final parameters
- RCIG can be seen as a generalization of FRePo, as changing Platt loss to MSE-loss

![image](https://user-images.githubusercontent.com/19633934/231650691-57820cb7-7af1-4bbb-9316-140ba86150df.png)

# Results
![image](https://user-images.githubusercontent.com/19633934/231644195-db99988f-223b-4852-b57c-0f53d5a6c120.png)

## TESLA results
![image](https://user-images.githubusercontent.com/19633934/231644270-4dfcbf87-9117-4a32-a673-d69916001eb5.png)
